File: src/router/repository/ranking.ts
import { type Origin } from '@prisma/client'
import type { NextFunction, Request, Response } from '@tinyhttp/app'
import { prisma } from 'database.js'

type SearchResponse = Response & {
	locals: {
		query: string;
	};
}

export function middleware(request: Request, response: SearchResponse, next: NextFunction) {
	const query = request.query.rank
	if (!query) {
		return response.status(400)
			.json({
				message: '400 Bad Request',
				error: 'Missing query parameter: \'rank\'',
				date: new Date()
			})
	}

	if (query.length !== 1) {
		return response.status(400)
			.json({
				message: '400 Bad Request',
				error: 'Query parameter \'q\' must be a single character of (1,2,3,4,5,*)',
				date: new Date()
			})
	}

	response.locals.query = query.toString()
	next()
}

export async function handler(request: Request, response: SearchResponse) {
	const { query } = response.locals

	const repos = query === '*' ? await prisma.repository.findMany({
		orderBy: {
			tier: 'asc'
		},

		include: {
			origin: true
		}
	}) : await prisma.repository.findMany({
		where: {
			tier: Number(query),
			isPruned: false
		},

		include: {
			origin: true
		},

		orderBy: {
			name: 'asc'
		}
	})

	return response.status(200)
		.json({
			message: '200 Successful',
			date: new Date(),
			count: repos.length,
			data: repos.map(data => {
				const entries = Object.entries(data)
					.map(([key, value]) => {
						if (key === 'origin') {
							const filtered = Object.fromEntries(Object.entries(value as Origin)
								.filter(([key]) => key !== 'uuid'))

							return [key, filtered]
						}

						return [key, value]
					})
					.filter(([key]) => key !== 'originId' && key !== 'isPruned')

				// eslint-disable-next-line @typescript-eslint/no-unsafe-return
				return {
					...Object.fromEntries(entries),
					refs: {
						meta: `${$product.api_endpoint}/jailbreak/repository/${data.slug}`,
						packages: `${$product.api_endpoint}/jailbreak/repository/${data.slug}/packages`
					}
				}
			})
		})
}

File: api/src/routes/repository/ranking.rs
use crate::utility::{merge_json, tokio_run};
use crate::{db::prisma, utility::json_respond};

use crate::prisma::repository;
use prisma_client_rust::Direction;
use serde_json::{json, Value};
use tide::{
	prelude::Deserialize,
	Request, Result,
	StatusCode::{BadRequest, Ok as OK, UnprocessableEntity},
};

#[derive(Deserialize)]
struct Query {
	rank: Option<String>,
}

pub async fn repository_ranking(req: Request<()>) -> Result {
	let query = match req.query::<Query>() {
		Ok(query) => {
			let rank = match query.rank {
				Some(q) => {
					let match_q = match q.as_str() {
						"1" => q,
						"2" => q,
						"3" => q,
						"4" => q,
						"5" => q,
						"*" => q,
						_ => {
							return Ok(json_respond(
								BadRequest,
								json!({
									"message": "400 Bad Request",
									"error": "Query parameter \'rank\' must be 1, 2, 3, 4, 5, or *",
									"date": chrono::Utc::now().to_rfc3339(),
								}),
							));
						}
					};

					match_q
				}
				None => {
					return Ok(json_respond(
						BadRequest,
						json!({
							"message": "400 Bad Request",
							"error": "Missing query parameter: \'rank\'",
							"date": chrono::Utc::now().to_rfc3339(),
						}),
					));
				}
			};

			rank
		}

		Err(err) => {
			println!("Error: {}", err);
			return Ok(json_respond(
				UnprocessableEntity,
				json!({
					"message": "422 Unprocessable Entity",
					"error": "Malformed query parameters",
					"date": chrono::Utc::now().to_rfc3339(),
				}),
			));
		}
	};

	let repositories = tokio_run(async move {
		return match query.as_str() {
			"*" => prisma()
				.await
				.repository()
				.find_many(vec![repository::is_pruned::equals(false)])
				.order_by(repository::tier::order(Direction::Asc))
				.with(repository::origin::fetch())
				.exec()
				.await
				.unwrap(),

			_ => prisma()
				.await
				.repository()
				.find_many(vec![
					repository::tier::equals(query.parse::<i32>().unwrap()),
					repository::is_pruned::equals(false),
				])
				.order_by(repository::tier::order(Direction::Asc))
				.with(repository::origin::fetch())
				.exec()
				.await
				.unwrap(),
		};
	});

	return Ok(json_respond(
		OK,
		json!({
			"message": "200 Successful",
			"date": chrono::Utc::now().to_rfc3339(),
			"count": repositories.len(),
			"data": repositories.iter().map(|repository|{
				let slug = repository.slug.clone();

				return merge_json(repository, json!({
					"refs": {
						"meta": format!("{}/jailbreak/repository/{}", env!("CANISTER_API_ENDPOINT"), slug),
						"packages": format!("{}/jailbreak/repository/{}/packages", env!("CANISTER_API_ENDPOINT"), slug),
					}
				}))
			}).collect::<Vec<Value>>(),
		}),
	));
}
